DOC: update the DataFrame.loc[] docstring

akosel opened the pull request:
Checklist for the pandas documentation sprint (ignore this if you are doing
an unrelated PR):

- [x] PR title is "DOC: update the <your-function-or-method> docstring"
- [x] The validation script passes: `scripts/validate_docstrings.py <your-function-or-method>`
- [x] The PEP8 style check passes: `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] The html version looks good: `python doc/make.py --single <your-function-or-method>`
- [x] It has been proofread on language by another sprint participant

Please include the output of the validation script below between the "```" ticks:

```
# paste output of "scripts/validate_docstrings.py <your-function-or-method>" here
# between the "```" (remove this comment, but keep the "```")
################################################################################
####################### Docstring (pandas.DataFrame.loc) #######################
################################################################################

Access a group of rows and columns by label(s) or a boolean array.

``.loc[]`` is primarily label based, but may also be used with a
boolean array.

Allowed inputs are:

- A single label, e.g. ``5`` or ``'a'``, (note that ``5`` is
  interpreted as a *label* of the index, and **never** as an
  integer position along the index).
- A list or array of labels, e.g. ``['a', 'b', 'c']``.
- A slice object with labels, e.g. ``'a':'f'``.

.. warning:: Note that contrary to usual python slices, **both** the start
    and the stop are included

- A boolean array of the same length as the axis being sliced,
  e.g. ``[True, False, True]``.
- A ``callable`` function with one argument (the calling Series, DataFrame
  or Panel) and that returns valid output for indexing (one of the above)

See more at :ref:`Selection by Label <indexing.label>`

See Also
--------
DateFrame.at : Access a single value for a row/column label pair
DateFrame.iloc : Access group of rows and columns by integer position(s)
DataFrame.xs : Returns a cross-section (row(s) or column(s)) from the
    Series/DataFrame.
Series.loc : Access group of values using labels

Examples
--------
**Getting values**

>>> df = pd.DataFrame([[1, 2], [4, 5], [7, 8]],
...      index=['cobra', 'viper', 'sidewinder'],
...      columns=['max_speed', 'shield'])
>>> df
            max_speed  shield
cobra               1       2
viper               4       5
sidewinder          7       8

Single label. Note this returns the row as a Series.

>>> df.loc['viper']
max_speed    4
shield       5
Name: viper, dtype: int64

List of labels. Note using ``[[]]`` returns a DataFrame.

>>> df.loc[['viper', 'sidewinder']]
            max_speed  shield
viper               4       5
sidewinder          7       8

Single label for row and column

>>> df.loc['cobra', 'shield']
2

Slice with labels for row and single label for column. As mentioned
above, note that both the start and stop of the slice are included.

>>> df.loc['cobra':'viper', 'max_speed']
cobra    1
viper    4
Name: max_speed, dtype: int64

Boolean list with the same length as the row axis

>>> df.loc[[False, False, True]]
            max_speed  shield
sidewinder          7       8

Conditional that returns a boolean Series

>>> df.loc[df['shield'] > 6]
            max_speed  shield
sidewinder          7       8

Conditional that returns a boolean Series with column labels specified

>>> df.loc[df['shield'] > 6, ['max_speed']]
            max_speed
sidewinder          7

Callable that returns a boolean Series

>>> df.loc[lambda df: df['shield'] == 8]
            max_speed  shield
sidewinder          7       8

**Setting values**

Set value for all items matching the list of labels

>>> df.loc[['viper', 'sidewinder'], ['shield']] = 50
>>> df
            max_speed  shield
cobra               1       2
viper               4      50
sidewinder          7      50

Set value for an entire row

>>> df.loc['cobra'] = 10
>>> df
            max_speed  shield
cobra              10      10
viper               4      50
sidewinder          7      50

Set value for an entire column

>>> df.loc[:, 'max_speed'] = 30
>>> df
            max_speed  shield
cobra              30      10
viper              30      50
sidewinder         30      50

Set value for rows matching callable condition

>>> df.loc[df['shield'] > 35] = 0
>>> df
            max_speed  shield
cobra              30      10
viper               0       0
sidewinder          0       0

**Getting values on a DataFrame with an index that has integer labels**

Another example using integers for the index

>>> df = pd.DataFrame([[1, 2], [4, 5], [7, 8]],
...      index=[7, 8, 9], columns=['max_speed', 'shield'])
>>> df
   max_speed  shield
7          1       2
8          4       5
9          7       8

Slice with integer labels for rows. As mentioned above, note that both
the start and stop of the slice are included.

>>> df.loc[7:9]
   max_speed  shield
7          1       2
8          4       5
9          7       8

**Getting values with a MultiIndex**

A number of examples using a DataFrame with a MultiIndex

>>> tuples = [
...    ('cobra', 'mark i'), ('cobra', 'mark ii'),
...    ('sidewinder', 'mark i'), ('sidewinder', 'mark ii'),
...    ('viper', 'mark ii'), ('viper', 'mark iii')
... ]
>>> index = pd.MultiIndex.from_tuples(tuples)
>>> values = [[12, 2], [0, 4], [10, 20],
...         [1, 4], [7, 1], [16, 36]]
>>> df = pd.DataFrame(values, columns=['max_speed', 'shield'], index=index)
>>> df
                     max_speed  shield
cobra      mark i           12       2
           mark ii           0       4
sidewinder mark i           10      20
           mark ii           1       4
viper      mark ii           7       1
           mark iii         16      36

Single label. Note this returns a DataFrame with a single index.

>>> df.loc['cobra']
         max_speed  shield
mark i          12       2
mark ii          0       4

Single index tuple. Note this returns a Series.

>>> df.loc[('cobra', 'mark ii')]
max_speed    0
shield       4
Name: (cobra, mark ii), dtype: int64

Single label for row and column. Similar to passing in a tuple, this
returns a Series.

>>> df.loc['cobra', 'mark i']
max_speed    12
shield        2
Name: (cobra, mark i), dtype: int64

Single tuple. Note using ``[[]]`` returns a DataFrame.

>>> df.loc[[('cobra', 'mark ii')]]
               max_speed  shield
cobra mark ii          0       4

Single tuple for the index with a single label for the column

>>> df.loc[('cobra', 'mark i'), 'shield']
2

Slice from index tuple to single label

>>> df.loc[('cobra', 'mark i'):'viper']
                     max_speed  shield
cobra      mark i           12       2
           mark ii           0       4
sidewinder mark i           10      20
           mark ii           1       4
viper      mark ii           7       1
           mark iii         16      36

Slice from index tuple to index tuple

>>> df.loc[('cobra', 'mark i'):('viper', 'mark ii')]
                    max_speed  shield
cobra      mark i          12       2
           mark ii          0       4
sidewinder mark i          10      20
           mark ii          1       4
viper      mark ii          7       1

Raises
------
KeyError:
    when any items are not found

################################################################################
################################## Validation ##################################
################################################################################

Errors found:
	No returns section found

```

If the validation script still gives errors, but you think there is a good reason
to deviate in this case (and there are certainly such cases), please state this
explicitly.

akosel committed:
DOC: update the DataFrame.loc[] docstring

TomAugspurger commented on pandas/core/indexing.py:
```diff
@@ -1413,10 +1413,13 @@ def _get_slice_axis(self, slice_obj, axis=None):
 
     ``.loc[]`` is primarily label based, but may also be used with a
-    boolean array.
+    boolean array. Note that if no row or column labels are specified
```
I don't think this sentence with "Note that..." is correct. I woudl remove it.

TomAugspurger commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1429,53 @@ class _LocIndexer(_LocationIndexer):
     - A slice object with labels, e.g. ``'a':'f'`` (note that contrary
       to usual python slices, **both** the start and the stop are included!).
-    - A boolean array.
+    - A boolean array, e.g. [True, False, True].
```
specify that this should be the same length as the axis being sliced.

TomAugspurger commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1429,53 @@ class _LocIndexer(_LocationIndexer):
 
+    See Also
+    --------
+    at : Selects a single value for a row/column label pair
```
`DataFrame.` for all these.

TomAugspurger commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1429,53 @@ class _LocIndexer(_LocationIndexer):
+    r2  10  20  30
+    >>> df.loc[df['c1'] > 10, ['c0', 'c2']]
+        c0  c2
+    r2  10  30
```
Can you make a second example series or DataFrame where the index values are integers, but not `0-len(df)`? And then show how `.loc` uses the *labels* and not the positions?

In that second example, could you also show a slice like `df.loc[2:5]` and show that it's closed on the right, so the label 5 is included?

TomAugspurger commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1429,53 @@ class _LocIndexer(_LocationIndexer):
+    r0    12
+    r1     0
+    Name: c0, dtype: int64
+    >>> df.loc[[False, False, True]]
```
Would be nice to have small bits of text breaking these up. Like "Indexing with a boolean array."

jreback commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1429,53 @@ class _LocIndexer(_LocationIndexer):
+    r0  12   2   3
+    r1   0   4   1
+    r2  10  20  30
+    >>> df.loc['r1']
```
blank lines in between cases

akosel committed:
More labels for examples. More examples. Update wording based on PR comments

akosel commented:
Thank you for the comments! I added commentary on each of the different examples. I also added another example `df` with integers set for the index labels. Please let me know if you have any other changes you'd like to see.

jorisvandenbossche commented on pandas/core/indexing.py:
```diff
@@ -1413,7 +1413,8 @@ def _get_slice_axis(self, slice_obj, axis=None):
 class _LocIndexer(_LocationIndexer):
-    """Purely label-location based indexer for selection by label.
+    """
+    Selects a group of rows and columns by label(s) or a boolean array.
```
"Select" is not fully correct as it is not only for getting, but also for setting? 
Or is that general enough? (@jreback @TomAugspurger ) To set you of course also need to select the location to set .. 
The "see also" is using "access" now instead of "select"

akosel replied:
Ah yes, I had meant to switch to using `access` all around, although I'm not sure if that gets around the problem you mentioned. Perhaps, it's enough to mention you can use `loc` to get and set in the extended summary? I'm struggling to think of a good word that implies both getting and setting...but I will keep thinking on it.

akosel replied:
And I can also add some examples of using `loc` for setting values below.

jorisvandenbossche replied:
yes, that would be good

jorisvandenbossche commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1427,105 @@ class _LocIndexer(_LocationIndexer):
+    c2    1
+    Name: r1, dtype: int64
+
+
```
only a single blank line (below as well)

akosel committed:
Remove lines and be consistent with wording

akosel committed:
Add examples of setting values with loc

jreback commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1427,140 @@ class _LocIndexer(_LocationIndexer):
 
+    See Also
+    --------
+    DateFrame.at
```
add Series.loc

akosel replied:
Added

jreback commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1427,140 @@ class _LocIndexer(_LocationIndexer):
+    r1   0   4   1
+    r2  10  20  30
+
+    Single label for row (note it would be faster to use ``DateFrame.at`` in
```
don't need the note about perf

jreback replied:
you can mention that this returns a Series and using ``[[]]`` returns a DataFrame (or maybe just show that as an example right after)

jreback commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1427,140 @@ class _LocIndexer(_LocationIndexer):
+    c2    1
+    Name: r1, dtype: int64
+
+    Single label for row and column (note it would be faster to use
```
same, remove the commentary

jreback commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1427,140 @@ class _LocIndexer(_LocationIndexer):
+    >>> df = pd.DataFrame([[12, 2, 3], [0, 4, 1], [10, 20, 30]],
+    ...      index=[7, 8, 9], columns=['c0', 'c1', 'c2'])
+    >>> df
+        c0  c1  c2
```
nice examples! can you add one using a MultiIndex for the index, and show selecting with tuples. sure this is getting long, but these examples are useful.

jreback commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1427,140 @@ class _LocIndexer(_LocationIndexer):
+    --------
+    DateFrame.at
+        Access a single value for a row/column label pair
+    DateFrame.iat
```
you can remove .iat from here (leave the .iloc though)

akosel committed:
Update See Also. Add more examples and specifies return type

akosel commented:
Thank you again for the comments. I added a number of other examples with DataFrame that has a MultiIndex. Let me know if it was overkill, or if I missed an important use case (or included one that is not very useful).

jorisvandenbossche commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1427,220 @@ class _LocIndexer(_LocationIndexer):
+    DateFrame.at
+        Access a single value for a row/column label pair
+    DateFrame.iloc
+        Access group of rows and columns by integer position(s)
```
Same comment here as on the other PR: 

```
DateFrame.iloc : explanation ..
```

jorisvandenbossche commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1427,220 @@ class _LocIndexer(_LocationIndexer):
+    r1   0   4   1
+    r2  10  20  30
+
+    Single label. Note this returns a Series.
```
Maybe clarify "Note this returns a Series." even further as "Note this returns the row as a Series."

jorisvandenbossche commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1427,220 @@ class _LocIndexer(_LocationIndexer):
+
+    Examples
+    --------
+    >>> df = pd.DataFrame([[12, 2, 3], [0, 4, 1], [10, 20, 30]],
```
I would maybe number the values consectively, so in the output it is easier to see which row was returned

jorisvandenbossche commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1427,220 @@ class _LocIndexer(_LocationIndexer):
+
+    Callable that returns valid output for indexing
+
+    >>> df.loc[df['c1'] > 10]
```
This is actually not a callable but a boolean series, same for the example below. I think this is a nice example to keep though, but would explain it a bit different (frame it as a boolean Series that is calculated from the frame itself)

jorisvandenbossche reviewed (commented):
Some great examples you added! Given that it are quite some, can you add some subsection titles? (you can use bold for that (`**Sub section**`) as done in the docstring guide). 
For example subsection for getting, setting, MultiIndex (you can see a bit how it can be divided, it is meant to give the long Examples section a bit more structure)

akosel committed:
Better labeling of subsections within docs

akosel commented:
@jorisvandenbossche Thank you for the recommendation on using subsections within the examples--I think that's a great idea. Please let me know if you think there is anything else I should add or modify.

jreback commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1427,231 @@ class _LocIndexer(_LocationIndexer):
+    c2    6
+    Name: r1, dtype: int64
+
+    List with a single label. Note using ``[[]]`` returns a DataFrame.
```
this is slightly redudant as you are showing the example with a list below

jreback commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1427,231 @@ class _LocIndexer(_LocationIndexer):
+
+    Boolean list with the same length as the row axis
+
+    >>> df.loc[[False, False, True]]
```
this is not a very common thing to do (directly), the boolean indexing right below is MUCH more important.

jreback commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1427,231 @@ class _LocIndexer(_LocationIndexer):
+
+    Boolean list
+
+    >>> df.loc[[True, False, True, False, True, True]]
```
same as above (remove this example)

jreback commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1427,231 @@ class _LocIndexer(_LocationIndexer):
+    Raises
+    ------
+    KeyError:
+        when items are not found
```
when *any* items are not found

jreback reviewed (changes requested):
just a few comments. looks pretty good!. @jorisvandenbossche @TomAugspurger @datapythonista

datapythonista commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1427,231 @@ class _LocIndexer(_LocationIndexer):
     - A list or array of labels, e.g. ``['a', 'b', 'c']``.
     - A slice object with labels, e.g. ``'a':'f'`` (note that contrary
       to usual python slices, **both** the start and the stop are included!).
```
May be we could use the `.. warning::` directive for this comment (instead of a note in brackets ended with the exclamation mark).

datapythonista reviewed (commented):
Great job for a quite difficult class to document. Added couple of comments with ideas.

akosel committed:
Update based on feedback

akosel commented on pandas/core/indexing.py:
```diff
@@ -1426,14 +1427,231 @@ class _LocIndexer(_LocationIndexer):
+    **Getting values**
+
+    >>> df = pd.DataFrame([[1, 2, 3], [4, 5, 6], [7, 8, 9]],
+    ...      index=['r0', 'r1', 'r2'], columns=['c0', 'c1', 'c2'])
```
I agree. I have added some labels some more meaningful labels. Let me know if you like it or have any other feedback on this matter.

datapythonista reviewed (approved):
lgtm, really clear and useful examples.

TomAugspurger commented:
This looks really good.

Question: do we want to mention / document indexing with duplicates? Having duplicates in the index is such a mess that I'm fine with avoiding it, but curious what others think.

jorisvandenbossche committed:
rst formatting

jorisvandenbossche commented:
> Question: do we want to mention / document indexing with duplicates? Having duplicates in the index is such a mess that I'm fine with avoiding it, but curious what others think

I think we need to discuss a bit to what extent the docstring of methods should be the full reference, or more the basic usage and having the full reference with the many examples in the reference guide (as it is now). But that's a more general discussion, as then we should also need a better way to structure the docstrings with subsections I think. 

So given that, I would leave this docstring now as it is (keeping it to basic usage). As starting to discuss duplicates as well, it will get much longer.

TomAugspurger commented:
👍 

Thanks all!

akosel commented:
Thank you all once again!
